[SIL] Maintain owned argument lifetimes at inlining.

nate-chandler opened the pull request:
Previously, `begin_borrow [lexical]` were created during SILGen for @owned arguments.  Such borrows could be deleted if trivially dead, which was the original reason why @owned arguments were considered lexical and could not have their destroys hoisted.

Those borrows were however important during inlining because they would maintain the lifetime of the owned argument.  Unless of course the borrow scope was trivially dead.  In which case the owned argument's lifetime would not be maintained.  And if the caller's value was non-lexical, destroys of the value could be hoisted over deinit barriers.

Here, during inlining, `move_value [lexical]`s are introduced during inlining whever the caller's value is non-lexical.  This maintains the lifetime of the owned argument even after inlining.

Additionally, took the opportunity to improve the handling of guaranteed function argument lifetimes: (1) during inlining they don't get lexical borrow_scopes when the caller's value is already lexical and (2) they are recongized as lexical.

nate-chandler committed:
[ObjectOutliner] See through move_value insts.

Enables the outlining of the ContiguousArrayStorage<StaticString> used
when initializing a RawRepresentable enum whose RawValue is String into
a global value to continue even when ContiguousArrayStorage has a
lexical lifetime.

Addresses the following regressions

StringEnumRawValueInitialization    400     7680    +1820.0%   **0.05x**
ArrayLiteral2                       78      647     +729.5%    **0.12x**
DataCreateSmallArray                1750    8850    +405.7%    **0.20x**

seen when enabling lexical lifetimes in the standard library.

nate-chandler committed:
[OwnershipUtils] Match complex nested borrows.

Recognize lexical borrows as nested when their borrowee's guaranteed
reference roots are all lexical borrows.

Addresses the following regressions

Breadcrumbs.MutatedUTF16ToIdx.Mixed         188     882     +369.1%   **0.21x**
Breadcrumbs.MutatedIdxToUTF16.Mixed         230     926     +302.6%   **0.25x**

seen when enabling lexical lifetimes in the standard library.

nate-chandler committed:
[WalkUtils] Walk through move_value instructions.

Addresses the following regressions

StackPromo                                10100   14400   +42.6%   **0.70x**

seen when enabling lexical lifetimes in the standard library.

nate-chandler committed:
[SIL] All lexical arguments are lexical.

Previously, only @owned function arguments were treated as lexical.
Here, all function arguments which report themselves to be lexical (i.e.
based on their type and annotation) are regarded as lexical.

nate-chandler committed:
[SILInliner] Only add lexical borrows when needed.

If a guaranted function argument is lexical, its lifetime must be
mainted during inlining.  If the caller's value is already lexical,
no work is required to do the inlining.

nate-chandler committed:
[SIL] Protect owned arg lifetimes at inlining.

Previously, `begin_borrow [lexical]` were created during SILGen for
@owned arguments.  Such borrows could be deleted if trivially dead,
which was the original reason why @owned arguments were considered
lexical and could not have their destroys hoisted.

Those borrows were however important during inlining because they would
maintain the lifetime of the owned argument.  Unless of course the
borrow scope was trivially dead.  In which case the owned argument's
lifetime would not be maintained.  And if the caller's value was
non-lexical, destroys of the value could be hoisted over deinit
barriers.

Here, during inlining, `move_value [lexical]`s are introduced during
inlining whever the caller's value is non-lexical.  This maintains the
lifetime of the owned argument even after inlining.

nate-chandler committed:
[Test] Adjusted SIL tests.

Now that
- owned arguments don't get lexical borrow scopes from SILGen
- owned arguments get lexical moves during inlining unless the value
  passed for that argument is already lexical
- guaranteed arguments don't get lexical borrow scopes during inlining
  when the value passed for that argument is already lexical
tests involving owned argument emission from SILGen and inlining need to
be updated.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
```
REGRESSION                             OLD    NEW    DELTA    RATIO
MapReduceClass2                        10     12     +20.0%   **0.83x (?)**
NormalizedIterator_slowerPrenormal     280    310    +10.7%   **0.90x (?)**
NormalizedIterator_fastPrenormal       470    520    +10.6%   **0.90x (?)**
Set.isSubset.Seq.Int25                 47     51     +8.5%    **0.92x (?)**

IMPROVEMENT                            OLD    NEW    DELTA    RATIO
FlattenListFlatMap                     4490   2582   -42.5%   **1.74x (?)**
Set.isDisjoint.Int100                  211    131    -37.9%   **1.61x**
Set.isDisjoint.Int50                   208    131    -37.0%   **1.59x**
Set.isDisjoint.Int25                   208    131    -37.0%   **1.59x**
SetIsSubsetInt0                        211    133    -37.0%   **1.59x**
FlattenListLoop                        1623   1066   -34.3%   **1.52x (?)**
Set.isSubset.Empty.Int                 126    98     -22.2%   **1.29x (?)**
Set.isStrictSubset.Empty.Int           128    100    -21.9%   **1.28x**
StringInterpolationManySmallSegments   8700   7400   -14.9%   **1.18x (?)**
Breadcrumbs.MutatedUTF16ToIdx.Mixed    196    176    -10.2%   **1.11x (?)**
Breadcrumbs.MutatedIdxToUTF16.Mixed    198    180    -9.1%    **1.10x (?)**
Diffing.Similar                        377    347    -8.0%    **1.09x (?)**
MapReduceLazyCollectionShort           30     28     -6.7%    **1.07x (?)**
```

nate-chandler committed:
[InstUtils] Look through move as ownership inst.

Add move_value to the list of ownership instructions through which
lookThroughOwnershipInsts looks.

nate-chandler committed:
[MandatoryInlining] Strip all ownership insts.

Previously, just copies and borrows were stripped.  All ownership
instructions, specifically, newly, move_value instructions, should be
stripped.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler committed:
[InstUtils] Look through move as copy/cast inst.

Add move_value to the list of instructions through which
getSingleValueCopyOrCast looks.

nate-chandler committed:
[ConstExpr] Look through move_value insts.

Just like begin_borrow and copy_value, move_value should be looked
through because the interpreter doesn't model the memory operation it
encodes.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci Please ASAN test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please clean benchmark

nate-chandler committed:
[Projection] Look through move_value insts.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
```
------- Performance (x86_64): -O -------

REGRESSION                                  OLD       NEW       DELTA     RATIO
ArrayAppendGenericStructs                   593.333   1260.0    +112.4%   **0.47x (?)**
Data.append.Sequence.64kB.Count             2.0       2.667     +33.3%    **0.75x (?)**
MapReduceClass2                             10.663    12.506    +17.3%    **0.85x (?)**
FlattenListFlatMap                          3926.0    4495.0    +14.5%    **0.87x (?)**

IMPROVEMENT                                 OLD       NEW       DELTA     RATIO
Set.isDisjoint.Int100                       208.083   129.611   -37.7%    **1.61x**
SetIsSubsetInt0                             206.3     129.133   -37.4%    **1.60x**
Set.isDisjoint.Int25                        204.1     129.667   -36.5%    **1.57x**
Set.isDisjoint.Int50                        202.6     129.733   -36.0%    **1.56x**
Set.isSubset.Empty.Int                      123.067   95.867    -22.1%    **1.28x**
Set.isStrictSubset.Empty.Int                125.231   98.0      -21.7%    **1.28x (?)**
Data.hash.Medium                            29.167    24.477    -16.1%    **1.19x (?)**
FlattenListLoop                             1616.0    1386.0    -14.2%    **1.17x (?)**
NormalizedIterator_nonBMPSlowestPrenormal   446.607   383.542   -14.1%    **1.16x (?)**
NormalizedIterator_emoji                    373.12    322.143   -13.7%    **1.16x (?)**
NormalizedIterator_slowerPrenormal          298.571   271.186   -9.2%     **1.10x (?)**
Set.isStrictSubset.Int.Empty                43.846    40.286    -8.1%     **1.09x (?)**
------- Performance (x86_64): -Osize -------

REGRESSION                       OLD         NEW         DELTA    RATIO
SetSymmetricDifferenceInt100     105.929     145.273     +37.1%   **0.73x**
SetSymmetricDifferenceInt50      120.385     146.75      +21.9%   **0.82x (?)**
FlattenListLoop                  1386.0      1613.0      +16.4%   **0.86x (?)**
SetSymmetricDifferenceInt25      121.333     139.154     +14.7%   **0.87x (?)**
MapReduceString                  42.333      46.88       +10.7%   **0.90x (?)**
PrefixAnySeqCRangeIterLazy       121.4       133.9       +10.3%   **0.91x (?)**

IMPROVEMENT                      OLD         NEW         DELTA    RATIO
Set.isDisjoint.Int100            208.0       129.611     -37.7%   **1.60x**
Set.isDisjoint.Int25             208.0       129.615     -37.7%   **1.60x**
Set.isDisjoint.Int50             202.6       129.667     -36.0%   **1.56x**
SetIsSubsetInt0                  204.8       132.933     -35.1%   **1.54x**
Dictionary4                      234.857     158.75      -32.4%   **1.48x (?)**
Dictionary4OfObjects             271.833     202.333     -25.6%   **1.34x (?)**
Set.isStrictSubset.Empty.Int     126.333     98.0        -22.4%   **1.29x (?)**
Set.isSubset.Empty.Int           123.067     95.824      -22.1%   **1.28x (?)**
StringWalk                       1530.435    1253.793    -18.1%   **1.22x**
SequenceAlgosArray               2511.111    2076.364    -17.3%   **1.21x (?)**
Set.filter.Int100.16k            24.979      20.963      -16.1%   **1.19x (?)**
Data.hash.Medium                 28.727      24.385      -15.1%   **1.18x (?)**
ObjectiveCBridgeStubFromNSDate   3255.0      2790.0      -14.3%   **1.17x (?)**
Set.filter.Int100.28k            42.885      36.79       -14.2%   **1.17x (?)**
Set.filter.Int100.20k            29.653      25.837      -12.9%   **1.15x (?)**
Chars2                           3580.952    3152.273    -12.0%   **1.14x (?)**
Set.filter.Int100.24k            34.5        30.907      -10.4%   **1.12x (?)**
Calculator                       152.846     136.933     -10.4%   **1.12x (?)**
NormalizedIterator_emoji         356.741     321.778     -9.8%    **1.11x (?)**
StringInterpolation              6536.842    5924.242    -9.4%    **1.10x (?)**
NopDeinit                        32683.333   29714.286   -9.1%    **1.10x (?)**

------- Code size: -Osize -------

REGRESSION                     OLD     NEW     DELTA   RATIO
NopDeinit.o                    4219    4401    +4.3%   **0.96x**
ClassArrayGetter.o             3629    3739    +3.0%   **0.97x**
DictionaryGroup.o              10896   11059   +1.5%   **0.99x**
SortLettersInPlace.o           8014    8122    +1.3%   **0.99x**
COWTree.o                      11141   11270   +1.2%   **0.99x**
DictionarySubscriptDefault.o   13418   13566   +1.1%   **0.99x**

IMPROVEMENT                    OLD     NEW     DELTA   RATIO
DictionaryCompactMapValues.o   13370   13155   -1.6%   **1.02x**
MapReduce.o                    20180   19943   -1.2%   **1.01x**
------- Performance (x86_64): -Onone -------

REGRESSION                   OLD       NEW       DELTA    RATIO
SubstringRemoveFirst1        0.143     0.167     +16.7%   **0.86x (?)**
CharacterLiteralsLarge       400.2     448.2     +12.0%   **0.89x (?)**

IMPROVEMENT                  OLD       NEW       DELTA    RATIO
Data.hash.Medium             33.125    28.824    -13.0%   **1.15x**
ArrayAppendUTF16Substring    25788.0   22764.0   -11.7%   **1.13x (?)**
ArrayAppendAsciiSubstring    25752.0   22764.0   -11.6%   **1.13x**
ArrayAppendLatin1Substring   26118.0   23088.0   -11.6%   **1.13x (?)**
ProtocolDispatch             506.0     457.5     -9.6%    **1.11x (?)**
SetIsSubsetInt100            448.0     412.167   -8.0%    **1.09x (?)**
SetIsSubsetInt25             136.294   125.571   -7.9%    **1.09x (?)**
Set.isStrictSubset.Int50     260.778   241.5     -7.4%    **1.08x (?)**
Set.isStrictSubset.Int25     136.588   127.111   -6.9%    **1.07x (?)**

------- Code size: -swiftlibs -------

```

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[OME] Lower move_value to a noop.

nate-chandler commented:
@swift-ci please benchmark

atrick commented on docs/SIL.rst:
```diff
@@ -2422,19 +2422,17 @@ A few sorts of SIL value have lifetimes that are constrained that way:
+While both @guaranteed and @owned function arguments are lexical, this is only
+interesting in the case of @owned arguments: @guaranteed function arguments are
+already guaranteed by the function's caller to live for the full duration of
+the function already.  Optimization of the function alone can't shorten it.  
```
At the language level, arguments are lexical under the same conditions as variables. Namely, that they don't have eager-move semantics, which includes CoW types and no-implicit-copy type qualifiers.

Note that when a user explicitly writes `borrow` or `consume`, the arguments are not lexical. Whether an argument is lexical is independent of the ownership convention.

nate-chandler replied:
Rewrote this.

nate-chandler commented:
```
------- Performance (x86_64): -O -------

REGRESSION                                  OLD       NEW       DELTA     RATIO
ArrayAppendGenericStructs                   623.333   1264.0    +102.8%   **0.49x (?)**
StringWithCString2                          0.0       0.001     +100.0%   **0.50x (?)**
NormalizedIterator_nonBMPSlowestPrenormal   391.017   449.02    +14.8%    **0.87x (?)**
NormalizedIterator_slowerPrenormal          274.699   301.5     +9.8%     **0.91x (?)**
NormalizedIterator_fastPrenormal            485.385   530.851   +9.4%     **0.91x (?)**
NormalizedIterator_emoji                    329.0     356.154   +8.3%     **0.92x (?)**

IMPROVEMENT                                 OLD       NEW       DELTA     RATIO
Set.isDisjoint.Seq.Int.Empty                53.512    47.192    -11.8%    **1.13x (?)**

------- Code size: -O -------
------- Performance (x86_64): -Osize -------

IMPROVEMENT                      OLD        NEW        DELTA    RATIO
Dictionary4                      232.0      158.167    -31.8%   **1.47x**
Dictionary4OfObjects             271.0      199.286    -26.5%   **1.36x**
Set.filter.Int100.20k            32.44      26.438     -18.5%   **1.23x (?)**
StringWalk                       1532.857   1251.765   -18.3%   **1.22x (?)**
ObjectiveCBridgeStubFromNSDate   3320.0     2865.714   -13.7%   **1.16x (?)**
Set.filter.Int100.28k            43.82      38.267     -12.7%   **1.15x**
FlattenListFlatMap               3427.0     3000.0     -12.5%   **1.14x (?)**
Set.filter.Int100.24k            35.258     31.45      -10.8%   **1.12x**
Set.filter.Int100.16k            24.12      21.528     -10.7%   **1.12x (?)**
StrComplexWalk                   3121.429   2797.143   -10.4%   **1.12x (?)**
Set.subtracting.Seq.Empty.Int    161.727    150.333    -7.0%    **1.08x (?)**

------- Code size: -Osize -------
------- Performance (x86_64): -Onone -------

REGRESSION                              OLD       NEW       DELTA    RATIO
ObjectiveCBridgeStubToNSStringRef       115.667   129.526   +12.0%   **0.89x (?)**
ObjectiveCBridgeStubNSDateRefAccess     2787.0    3092.0    +10.9%   **0.90x (?)**
CreateObjects                           1252.0    1361.0    +8.7%    **0.92x (?)**

IMPROVEMENT                             OLD       NEW       DELTA    RATIO
NSStringConversion.InlineBuffer.ASCII   5555.0    4925.0    -11.3%   **1.13x (?)**
NSStringConversion.InlineBuffer.UTF8    3501.0    3172.0    -9.4%    **1.10x (?)**
RC4                                     12397.0   11436.0   -7.8%    **1.08x (?)**

------- Code size: -swiftlibs -------

```

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci Please Apple Silicon benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please asan test

nate-chandler commented:
@swift-ci Please ASAN test

nate-chandler commented:
preset=asan
@swift-ci Please test with preset macOS

nate-chandler committed:
[CanonicalizeBorrowScope] Look through moves.

When encountering inside a borrow scope a non-lexical move_value or a
move_value [lexical] where the borrowed value is itself already lexical,
delete the move_value and regard its uses as uses of the moved-from
value.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please smoke test

nate-chandler committed:
[Test] Adjusted SIL tests.

Changes to teach the optimizer pipeline about move_value require some
corresponding test updates.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test macOS platform

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
```
------- Performance (x86_64): -O -------

REGRESSION           OLD      NEW      DELTA    RATIO
FlattenListFlatMap   2518.0   2787.0   +10.7%   **0.90x (?)**
Data.hash.Medium     26.567   28.731   +8.1%    **0.92x (?)**

------- Code size: -O -------

------- Performance (x86_64): -Osize -------

REGRESSION                     OLD      NEW        DELTA     RATIO
StringWithCString2             0.0      0.001      +100.0%   **0.50x (?)**
FlattenListFlatMap             3966.0   4377.0     +10.4%    **0.91x (?)**
StrComplexWalk                 2800.0   3037.143   +8.5%     **0.92x (?)**
StringFromLongWholeSubstring   2.717    2.935      +8.0%     **0.93x (?)**

IMPROVEMENT                    OLD      NEW        DELTA     RATIO
FlattenListLoop                1620.0   1386.0     -14.4%    **1.17x (?)**
DictionaryKeysContainsCocoa    17.0     15.46      -9.1%     **1.10x (?)**

------- Code size: -Osize -------

------- Performance (x86_64): -Onone -------

REGRESSION                     OLD     NEW     DELTA    RATIO
SetSymmetricDifferenceInt100   405.8   443.0   +9.2%    **0.92x (?)**

IMPROVEMENT                    OLD     NEW     DELTA    RATIO
CharacterLiteralsLarge         471.5   416.2   -11.7%   **1.13x (?)**

------- Code size: -swiftlibs -------
```

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please smoke test

atrick commented on lib/SILOptimizer/Utils/CanonicalizeBorrowScope.cpp:
```diff
@@ -224,21 +224,34 @@ bool CanonicalizeBorrowScope::visitBorrowScopeUses(SILValue innerValue,
-    for (Operand *use : uses) {
+    unsigned index = 0;
+    while (index < uses.size()) {
+      SWIFT_DEFER { ++index; };
```
a.k.a "for loop"?

atrick commented on lib/SILOptimizer/Utils/CanonicalizeBorrowScope.cpp:
```diff
@@ -224,21 +224,34 @@ bool CanonicalizeBorrowScope::visitBorrowScopeUses(SILValue innerValue,
+            uses.push_back(moveUse);
+          }
+          move->replaceAllUsesWith(move->getOperand());
+          deleter.forceDelete(move);
```
It's generally bad form to mutate SIL in the middle of an analysis--confusing/bug-prone from the client's perspective. Can't the analysis just scope the moves and let the rewriter visitors actually delete them?

nate-chandler replied:
Fixed to just add the move to the worklist.

nate-chandler committed:
[Test] Adapted test to different borrow scopes.

Borrow scopes were moved when lexical borrow scopes were no longer
emitted for owned arguments.  Now, they are emitted when guaranteed
values are needed.  Fixing these diagnostics will be tracked separately.
The same issue is already visible elsewhere in this test file, e.g. in
castTestSwitch1, castTestSwitch2, and castTestSwitchInLoop.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
```
------- Performance (x86_64): -O -------

REGRESSION                  OLD      NEW      DELTA    RATIO
Data.hash.Medium            26.778   28.964   +8.2%    **0.92x (?)**

IMPROVEMENT                 OLD      NEW      DELTA    RATIO
ArrayAppendGenericStructs   1092.0   580.0    -46.9%   **1.88x (?)**

------- Code size: -O -------

------- Performance (x86_64): -Osize -------

REGRESSION                     OLD      NEW        DELTA    RATIO
StringWalk                     1267.2   1394.118   +10.0%   **0.91x (?)**
StringFromLongWholeSubstring   2.606    2.826      +8.4%    **0.92x (?)**
Data.hash.Medium               26.556   28.733     +8.2%    **0.92x (?)**

IMPROVEMENT                    OLD      NEW        DELTA    RATIO
FlattenListLoop                1617.0   1240.0     -23.3%   **1.30x (?)**
FlattenListFlatMap             4389.0   3993.0     -9.0%    **1.10x (?)**
Data.hash.Empty                54.45    50.095     -8.0%    **1.09x (?)**

------- Code size: -Osize -------

------- Performance (x86_64): -Onone -------

REGRESSION                     OLD     NEW      DELTA     RATIO
ArrayAppendGenericStructs      610.0   1442.0   +136.4%   **0.42x (?)**
SetSymmetricDifferenceInt100   406.2   443.6    +9.2%     **0.92x (?)**

IMPROVEMENT                    OLD     NEW      DELTA     RATIO
CharacterLiteralsLarge         466.0   414.6    -11.0%    **1.12x (?)**

------- Code size: -swiftlibs -------

```

nate-chandler commented:
@swift-ci please test macOS platform

atrick reviewed (approved):
Looks great!

nate-chandler commented:
@swift-ci please test macOS platform
